src/FESpaces/Assemblers.jl

```diff
@@ -270,44 +270,88 @@ function assemble_matrix(f::Function,a::Assembler,U::FESpace,V::FESpace)
   assemble_matrix(a,collect_cell_matrix(U,V,f(u,v)))
 end
 
+function assemble_matrix!(f::Function,A::AbstractMatrix,a::Assembler,U::FESpace,V::FESpace)
+  v = get_fe_basis(V)
+  u = get_trial_fe_basis(U)
+  assemble_matrix!(A,a,collect_cell_matrix(U,V,f(u,v)))
+end
+
 function assemble_vector(f::Function,a::Assembler,V::FESpace)
   v = get_fe_basis(V)
   assemble_vector(a,collect_cell_vector(V,f(v)))
 end
 
+function assemble_vector!(f::Function,b::AbstractVector,a::Assembler,V::FESpace)
+  v = get_fe_basis(V)
+  assemble_vector!(b,a,collect_cell_vector(V,f(v)))
+end
+
 function assemble_matrix_and_vector(f::Function,b::Function,a::Assembler,U::FESpace,V::FESpace)
   v = get_fe_basis(V)
   u = get_trial_fe_basis(U)
   assemble_matrix_and_vector(a,collect_cell_matrix_and_vector(U,V,f(u,v),b(v)))
 end
 
+function assemble_matrix_and_vector!(f::Function,b::Function,M::AbstractMatrix,r::AbstractVector,a::Assembler,U::FESpace,V::FESpace)
+  v = get_fe_basis(V)
+  u = get_trial_fe_basis(U)
+  assemble_matrix_and_vector!(M,r,a,collect_cell_matrix_and_vector(U,V,f(u,v),b(v)))
+end
+
 function assemble_matrix(f,a::Assembler,U::FESpace,V::FESpace)
   assemble_matrix(a,collect_cell_matrix(U,V,f))
 end
 
+function assemble_matrix!(A::AbstractMatrix,f,a::Assembler,U::FESpace,V::FESpace)
+  assemble_matrix!(A,a,collect_cell_matrix(U,V,f))
+end
+
 function assemble_vector(f,a::Assembler,V::FESpace)
   assemble_vector(a,collect_cell_vector(V,f))
 end
 
+function assemble_vector!(b::AbstractVector,f,a::Assembler,V::FESpace)
```

Comment:
and here

Reply:
f is not a function, isnt?

Reply:
f is a callable object, usually a `Function`. 

to be consistent we have 2 main options:

- add  missing ::Function everywhere in this file, where we accept a function
- remove all ::Function anotations to allow generic callable objects

2 is more general, but 1 is less disruptive wrt what we have now and also works since the weak form is always defined via functions.

Reply:
I still dont understand why `f` it is function/callable object in this particular function. See, e.g., 

```
function collect_cell_vector(test::FESpace,a::DomainContribution)
  w = []
  r = []
  for trian in get_domains(a)
    cell_vec = get_contribution(a,trian)
    @assert ndims(eltype(cell_vec)) == 1
    cell_vec_r = attach_constraints_rows(test,cell_vec,trian)
    rows = get_cell_dof_ids(test,trian)
    push!(w,compress_contributions(cell_vec_r,trian))
    push!(r,compress_ids(rows,trian))
  end
  (w,r)
end
```

`a` is a callable object?

Reply:
You are right. It is not a function. it is an object for which `collect_cell_vector` and related functions are defined.

Reply:
ok, in any case I had to put the `f` and `b` args first in all function signatures to let all tests pass ...